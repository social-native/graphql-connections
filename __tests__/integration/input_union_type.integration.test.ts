import appProvider from '../appProvider';
import request from 'supertest';
import {createGQLRequest} from '../utils';

describe('Input Union Type', () => {
    const app = appProvider();

    describe('works with all unions', () => {
        it('not using variables', async () => {
            const gqlQuery = `
                query {
                    users(
                    filter:  { 
                        or: [
                        { field: "age", value: "41", operator: "="},
                        { field: "age", value: "31", operator: "="},
                        { and: [
                            { field: "haircolor", value: "gray", operator: "="},
                            { field: "age", value: "70", operator: "="},
                            { not: [
                            { field: "firstname", value: "Kenyon", operator: "="},
                                { field: "firstname", value: "Nerissa", operator: "="},
                            ]}
                        ]},
                        ],
                    }) {
                    pageInfo {
                        hasNextPage
                        hasPreviousPage
                    }
                    edges {
                        cursor
                        node {
                        id
                        age
                        haircolor
                        lastname
                        username
                        }
                    }
                    }
                }
            `;

            const response = await request(app.callback())
                .post('/graphql')
                .use(createGQLRequest(gqlQuery));

            expect(JSON.parse(response.text)).toHaveProperty(
                ['data', 'users', 'edges', 0, 'cursor'],
                expect.any(String)
            );
            expect(response.status).toBe(200);
        });

        it('using variables', async () => {
            const gqlQuery = `
                query($filter: Filter) {
                    users(filter: $filter) {
                    pageInfo {
                        hasNextPage
                        hasPreviousPage
                    }
                    edges {
                        cursor
                        node {
                        id
                        age
                        haircolor
                        lastname
                        username
                        }
                    }
                    }
                }
            `;

            const variables = `
            {
                "filter": { "or": { "field": "age", "value": "31", "operator": "=" }}
            }
            `;

            const response = await request(app.callback())
                .post('/graphql')
                .use(createGQLRequest(gqlQuery, variables));

            expect(JSON.parse(response.text)).toHaveProperty(
                ['data', 'users', 'edges', 0, 'cursor'],
                expect.any(String)
            );
            expect(response.status).toBe(200);
        });
    });

    describe('handles errors', () => {
        it('not using variables with compound filter', async () => {
            const gqlQuery = `
                query {
                    users(
                    filter:  {
                        or: [
                        { fields: "age", value: "41", operator: "="},
                        { field: "age", value: "31", operator: "="},
                        { and: [
                            { field: "haircolor", value: "gray", operator: "="},
                            { field: "age", value: "70", operator: "="},
                            { not: [
                            { field: "firstname", value: "Kenyon", operator: "="},
                                { field: "firstname", value: "Nerissa", operator: "="},
                            ]}
                        ]},
                        ],
                    }) {
                    pageInfo {
                        hasNextPage
                        hasPreviousPage
                    }
                    edges {
                        cursor
                        node {
                        id
                        age
                        haircolor
                        lastname
                        username
                        }
                    }
                    }
                }
            `;

            const response = await request(app.callback())
                .post('/graphql')
                .use(createGQLRequest(gqlQuery));

            expect(response.text).toMatchSnapshot();
            expect(response.status).toBe(400);
        });

        it('not using variables with single filter', async () => {
            const gqlQuery = `
                query {
                    users(
                        filter: { fields: "haircolor", value: "gray", operator: "=" }
                    ) {
                    pageInfo {
                        hasNextPage
                        hasPreviousPage
                    }
                    edges {
                        cursor
                        node {
                        id
                        age
                        haircolor
                        lastname
                        username
                        }
                    }
                    }
                }
            `;

            const response = await request(app.callback())
                .post('/graphql')
                .use(createGQLRequest(gqlQuery));

            expect(response.text).toMatchSnapshot();
            expect(response.status).toBe(400);
        });

        it('using variables with compound filter', async () => {
            const gqlQuery = `
                query($filter: FilterInputScalar) {
                    users(filter: $filter) {
                    pageInfo {
                        hasNextPage
                        hasPreviousPage
                    }
                    edges {
                        cursor
                        node {
                        id
                        age
                        haircolor
                        lastname
                        username
                        }
                    }
                    }
                }
            `;

            const variables = `
            {
                "filter": { "or": { "fields": "age", "value": "31", "operator": "=" }}
            }
            `;

            const response = await request(app.callback())
                .post('/graphql')
                .use(createGQLRequest(gqlQuery, variables));

            expect(response.text).toMatchSnapshot();
            expect(response.status).toBe(400);
        });

        it('using variables with single filter', async () => {
            const gqlQuery = `
                query($filter: FilterInputScalar) {
                    users(filter: $filter) {
                    pageInfo {
                        hasNextPage
                        hasPreviousPage
                    }
                    edges {
                        cursor
                        node {
                        id
                        age
                        haircolor
                        lastname
                        username
                        }
                    }
                    }
                }
            `;

            const variables = `
            {
                "filter": { "field": "age", "values": "30", "operator": "=" }
            }
            `;

            const response = await request(app.callback())
                .post('/graphql')
                .use(createGQLRequest(gqlQuery, variables));

            expect(response.text).toMatchSnapshot();
            expect(response.status).toBe(400);
        });
    });
});
