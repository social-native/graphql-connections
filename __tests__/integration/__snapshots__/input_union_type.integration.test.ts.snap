// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Union Type handles errors not using variables with compound filter 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Expected type Filter, found {or: [{fields: \\\\\\"age\\\\\\", value: \\\\\\"41\\\\\\", operator: \\\\\\"=\\\\\\"}, {field: \\\\\\"age\\\\\\", value: \\\\\\"31\\\\\\", operator: \\\\\\"=\\\\\\"}, {and: [{field: \\\\\\"haircolor\\\\\\", value: \\\\\\"gray\\\\\\", operator: \\\\\\"=\\\\\\"}, {field: \\\\\\"age\\\\\\", value: \\\\\\"70\\\\\\", operator: \\\\\\"=\\\\\\"}, {not: [{field: \\\\\\"firstname\\\\\\", value: \\\\\\"Kenyon\\\\\\", operator: \\\\\\"=\\\\\\"}, {field: \\\\\\"firstname\\\\\\", value: \\\\\\"Nerissa\\\\\\", operator: \\\\\\"=\\\\\\"}]}]}]}; Filter should be composed of either: CompoundFilterScalar \`{and: [Filter], or: [Filter], not: [Filter]}\`, or FilterScalar \`{field: String, operator: String, value: String}\`\\",\\"locations\\":[{\\"line\\":4,\\"column\\":30}],\\"extensions\\":{\\"code\\":\\"GRAPHQL_VALIDATION_FAILED\\"}}]}
"
`;

exports[`Input Union Type handles errors not using variables with single filter 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Expected type Filter, found {fields: \\\\\\"haircolor\\\\\\", value: \\\\\\"gray\\\\\\", operator: \\\\\\"=\\\\\\"}; Filter should be composed of either: CompoundFilterScalar \`{and: [Filter], or: [Filter], not: [Filter]}\`, or FilterScalar \`{field: String, operator: String, value: String}\`\\",\\"locations\\":[{\\"line\\":4,\\"column\\":33}],\\"extensions\\":{\\"code\\":\\"GRAPHQL_VALIDATION_FAILED\\"}}]}
"
`;

exports[`Input Union Type handles errors using variables with compound filter 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Unknown type \\\\\\"FilterInputScalar\\\\\\".\\",\\"locations\\":[{\\"line\\":2,\\"column\\":32}],\\"extensions\\":{\\"code\\":\\"GRAPHQL_VALIDATION_FAILED\\"}}]}
"
`;

exports[`Input Union Type handles errors using variables with single filter 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Unknown type \\\\\\"FilterInputScalar\\\\\\".\\",\\"locations\\":[{\\"line\\":2,\\"column\\":32}],\\"extensions\\":{\\"code\\":\\"GRAPHQL_VALIDATION_FAILED\\"}}]}
"
`;
